runner/organizations_test.go

```diff
@@ -22,19 +22,70 @@ import (
 	dbMocks "garm/database/common/mocks"
 	runnerErrors "garm/errors"
 	"garm/params"
+	"garm/runner/common"
+	poolManager "garm/runner/common/mocks"
+	providerMocks "garm/runner/common/mocks"
+	runnerMocks "garm/runner/mocks"
+	"strings"
 	"testing"
 
 	"github.com/stretchr/testify/require"
 )
 
+func TestCreateOrganization(t *testing.T) {
+	adminCtx := auth.GetAdminContext()
+	createOrgParams := params.CreateOrgParams{
+		Name:            "test",
+		CredentialsName: "test",
+		WebhookSecret:   "test",
+	}
+	testCreds := config.Github{
+		Name:        "test",
+		Description: "test",
+		OAuth2Token: "test-token",
+	}
+	createOrg := params.Organization{
+		ID:              "",
+		CredentialsName: "test",
+		WebhookSecret:   "",
+	}
+
+	providers := map[string]common.Provider{}
+
+	storeMock := dbMocks.NewStore(t)
+	poolMgrCtrlMock := runnerMocks.NewPoolManagerController(t)
+	poolMgrMock := poolManager.NewPoolManager(t)
+
+	storeMock.On("GetOrganization", adminCtx, createOrgParams.Name).Return(params.Organization{}, runnerErrors.ErrNotFound)
+	storeMock.On("CreateOrganization", adminCtx, createOrgParams.Name, testCreds.Name, createOrgParams.WebhookSecret).Return(params.Organization{CredentialsName: "test"}, nil)
+	poolMgrCtrlMock.On("CreateOrgPoolManager", adminCtx, createOrg, providers, storeMock).Return(poolMgrMock, nil)
+	poolMgrMock.On("Start").Return(nil)
+
+	runner := Runner{
+		credentials: map[string]config.Github{
+			"test": testCreds,
+		},
+		providers:       providers,
+		ctx:             adminCtx,
+		store:           storeMock,
+		poolManagerCtrl: poolMgrCtrlMock,
+	}
+
+	org, err := runner.CreateOrganization(adminCtx, createOrgParams)
+	storeMock.AssertExpectations(t)
+	poolMgrMock.AssertExpectations(t)
+	poolMgrCtrlMock.AssertExpectations(t)
+	require.Equal(t, createOrg, org)
```

Comment:
Comparing structs is a bit tricky. Especially if you have fields that point to other structs, or a slice of any kind. Slices are particularly tricky, because ordering matters when comparing. I am not sure if ```require.Equal()``` is enough in that case. For this canned result, I am certain you get the desired result, but it doesn't necessarily test the function properly. Estecially considering that the ```Organization{}``` structure also has a ```[]Pool``` field. It does test that we call the proper functions with expected args, which is good.

Reply:
You are right. I will update this as well.

Reply:
Can you clarify this further ? Should I avoid `require.Equal` for structs ? If so, what should I use to compare structs against expected structs ?

From other tests that I've written, I noticed that `require.Equal` works fine comparing structs (even if they contain slices, other structs, etc).

Reply:
For the test fixtures you use here, I am sure that comparing slices works just fine. However, it may fail when comparing unsorted slices. For example:

```go
require.Equal(t, [1,2,3], [3,2,1])
```

Even though the values are the same, the order differs. For cases where order matters, a failure is desirable. For cases where order does not matter, the error is a false positive. Also, this will most likely fail if one of the elements of a struct is a pointer. I am not sure if the ```require.Equal()``` walks through the struct and de-references pointers before comparing the value they point to. This is why we usually don't compare entire structs.

And in most cases, comparing the whole struct is not necessary. It is acceptable to compare values of individual fields we care about. In some cases, we only care that the value is non empty, as is the case of the auto-generated ```ID``` field in most objects returned by the API. We do care about fields such as ```Name``` in the case of an organization or repo. If we try to create an org and we request that it is called ```MyOrg``` then we expect the return value to have that name.